### PR TITLE
Add php 7 to travis

### PR DESCRIPTION
**Pull request by dimitri-koenig**
I reworked the travis config to make it all work. At least most of the time the tests run through, so the config is working.

The next step would be to fix failing tests and its source code.

### CODE REVIEW

**Comment by magento-cicd** [comment]
We have automated a Magento Contributor License Agreement verifier for contributions sent to our GitHub projects.
Please see the CLA agreement in the Pull Request comments below.

**Comment by dimitri-koenig** [comment]
Please merge this PR :-)

File: .travis.yml [hunk 1]
```diff
@@ -2,6 +2,8 @@ language: php
 php:
   - 5.5
   - 5.6
+  - 7.0
+  - hhvm
```
Comment (orlangur) [inline .travis.yml]: I don't see any test results produced by hhvm jobs, here is the example of such: https://travis-ci.org/Sanuch/magento2/builds/47961569

You can take needed changes from @Sanuch fork or remove hhvm occurrences from this PR.

File: .travis.yml [hunk 2]
```diff
@@ -15,12 +17,12 @@ cache:
     - $HOME/.composer/cache
     - $HOME/.cache/bin
 matrix:
-  exclude:
-    - php: 5.6
-      env: TEST_SUITE=static_phpcs
-    - php: 5.6
-      env: TEST_SUITE=static_annotation
```
Comment (orlangur) [inline .travis.yml]: This was made intentionally. In static tests application code is not executed thus there is no need to run them on different versions of PHP.

In such a way you would check that PHPCS works the same way under PHP5.6, PHP7 and other versions.

Please exclude all variants but PHP7 as the fastest one.

File: .travis.yml [hunk 3]
```diff
@@ -15,12 +17,12 @@ cache:
...
+  fast_finish: true
+  allow_failures:
+      - php: 7.0
+      - php: hhvm
 before_install:
+  - composer self-update
```
Comment (orlangur) [inline .travis.yml]: Why need this?

File: .travis.yml [hunk 4]
```diff
@@ -55,13 +52,21 @@ before_script:
       sudo apt-add-repository ppa:ondrej/mysql-5.6 -y;
       sudo apt-get update;
       sudo apt-get install mysql-server-5.6 mysql-client-5.6;
-      mysql -uroot -e 'SET @@global.sql_mode = NO_ENGINE_SUBSTITUTION; CREATE DATABASE magento_integration_tests;';
+      sudo mysql -e 'SET @@global.sql_mode = NO_ENGINE_SUBSTITUTION; CREATE DATABASE magento_integration_tests;';
```
Comment (orlangur) [inline .travis.yml]: Why need this?

File: .travis.yml [hunk 5]
```diff
@@ -55,13 +52,21 @@ before_script:
...
       mv dev/tests/integration/etc/install-config-mysql.travis.php.dist dev/tests/integration/etc/install-config-mysql.php;
       fi"
+
+  # try to avoid mysql has gone away errors
```
Comment (orlangur) [inline .travis.yml]: This try was not successful if I remember build results correctly.

Currently integration tests on mainline are green, so, please revert these changes.

File: .travis.yml [hunk 6]
```diff
@@ -55,13 +52,21 @@ before_script:
...
+  - echo -e "\nmax_allowed_packet=32M" | sudo tee -a /etc/mysql/my.cnf
+  - sudo service mysql restart
+  - sudo mysql -e "SHOW VARIABLES LIKE 'max_allowed_packet';"
+  - sudo mysql -e "SHOW VARIABLES LIKE 'wait_timeout';"
+
   # Change memory_limit for travis
   - echo 'memory_limit = -1' >> ~/.phpenv/versions/$(phpenv version-name)/etc/conf.d/travis.ini
   - phpenv rehash;
-  - composer install --no-interaction --prefer-dist
+  - composer install --no-interaction --prefer-dist --ignore-platform-reqs
```
Comment (orlangur) [inline .travis.yml]: This is for HHVM only which does not pass current `composer.json` restrictions?

**Comment by orlangur** [comment]
Very nice to see PHP7 jobs workable on Travis CI at last :+1:

**Comment by dimitri-koenig** [comment]
Did some cleanup and rework of this pr

**Comment by davidalger** [comment]
@dimitri-koenig Looking at the Travis CI report there are only 3 failures on the PHP 7 build, specifically unit tests verifying PHP readiness checks. They are failing on something related to the `always_populate_raw_post_data` php configuration. Think you could take a stab at resolving those 3 failures? It would be great to get tests running on PHP 7, and even better if they all passed. Looks like it's only one small step away. :)

**Comment by dimitri-koenig** [comment]
I really prefer to delegate this to a separate issue and so merge and close this pull request. One small change after another. Deal? :-)

**Comment by dimitri-koenig** [comment]
@davidalger Could you still merge this pull request?

**Comment by Vinai** [comment]
The github community team doesn't do merges, that is up to the core team. I think the chances for a merge would be greatly improved if the tests where all passing.
Even though I definitely can relate to favoring multiple smaller PRs, it's not our decision.

**Comment by adragus-inviqa** [comment]
@dimitri-koenig - I think it's just a simple matter of changing the expected arrays depending on the version of PHP and/or Hack lang, because some array keys in the PHP readiness check will be missing in PHP7/Hack: [link](https://github.com/magento/magento2/blob/3b6c4971dc08fee147e91f19a84a275772d79901/setup/src/Magento/Setup/Model/PhpReadinessCheck.php#L189-L192).

So, in `Magento\Setup\Test\Unit\Model\PhpReadinessCheckTest::testCheckPhpSettingsNoXDebug()` ([link](https://github.com/magento/magento2/blob/777f7828f01d6d1bac3a7924cacfc30df7a0aff0/setup/src/Magento/Setup/Test/Unit/Model/PhpReadinessCheckTest.php#L275-L281)), instead of 

``` php
$expected = [
    'responseType' => ResponseTypeInterface::RESPONSE_TYPE_SUCCESS,
    'data' => [
        'always_populate_raw_post_data' => [
            'message' => $rawPostMessage,
            'helpUrl' => 'http://php.net/manual/en/ini.core.php#ini.always-populate-settings-data',
            'error' => false
        ]
    ]
];
```

you'd do something like

``` php
/**
 * @return bool
 */
private function isPhp7OrHackLang()
{
    return version_compare(PHP_VERSION, '7.0.0-beta') >= 0 || defined('HHVM_VERSION');
}

// ...

$expected = [ 'responseType' => ResponseTypeInterface::RESPONSE_TYPE_SUCCESS ];

if (!$this->isPhp7OrHackLang()) {
    $expected['data'] = [
        'always_populate_raw_post_data' => [
            'message' => $rawPostMessage,
            'helpUrl' => 'http://php.net/manual/en/ini.core.php#ini.always-populate-settings-data',
            'error' => false
        ];
}
```

You'd have to do this 3 times, in the same class - `Magento\Setup\Test\Unit\Model\PhpReadinessCheckTest` - so you can reuse the `isPhp7OrHackLang()` and there you go - all PHP7 tests will pass. :tada:

**Comment by davidalger** [comment]
@adragus-inviqa Thanks for that. I implemented it in a branch on my own fork (see 7bca51c6ab). They still pass on 5.5 and 5.6 but give another strange error under PHP 5.7. Don't have time to try and solve it this week…but maybe next. If I do, and this PR isn't updated, I'll just submit one with the new changes included. If you've got suggestions, I'll take 'em too :)

```
There were 3 failures:

1) Magento\Setup\Test\Unit\Model\PhpReadinessCheckTest::testCheckPhpSettings
Expectation failed for method name is equal to <string:parseConstraints> when invoked at sequence index 0.
The expected invocation at index 0 was never reached.

2) Magento\Setup\Test\Unit\Model\PhpReadinessCheckTest::testCheckPhpSettingsFailed
Expectation failed for method name is equal to <string:parseConstraints> when invoked at sequence index 0.
The expected invocation at index 0 was never reached.

3) Magento\Setup\Test\Unit\Model\PhpReadinessCheckTest::testCheckPhpSettingsNoXDebug
Expectation failed for method name is equal to <string:parseConstraints> when invoked at sequence index 0.
The expected invocation at index 0 was never reached.
```

You can see the failing tests with new error messages here: https://travis-ci.org/davidalger/magento2/jobs/93043185

**Comment by adragus-inviqa** [comment]
Yeah, that's a head scratcher. Not sure what could've changed in 5.7, as that `to()` seems related to call order.
Or, as one of the comments in [here](http://www.andrejfarkas.com/2012/07/phpunit-at-method-to-check-method-invocation-at-certain-index/) put it: "expect method blah() to be called as the n-th method call on this mock...".

**Comment by davidalger** [comment]
@adragus-inviqa That link helped a bunch. There was a mock being setup within a method each of these tests called, but the mock isn't used when running PHP 7. Changed that and the tests all passed! I'm going to submit a PR with the entire set of changes now.
